### PR TITLE
Improve shape function of tf.sparse_reduce_sum

### LINKED ISSUE #23114: Can tf.sparse_reduce_sum infer shape while keepdims=True?
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.11
- Are you willing to contribute it (Yes/No): 
No. Sorry, I'm not familiar with the TF codebase.

**Describe the feature and the current behavior/state.**
tf.sparse_reduce_sum should be able to infer the shape if the input has known shape.
```pythonIn [1]: import tensorflow as tf

In [2]: x = tf.SparseTensor(indices=[[0,0],[1,1]], values=[1,2], dense_shape=[2,2])

In [3]: tf.sparse_reduce_sum(x)
Out[3]: <tf.Tensor 'SparseReduceSum:0' shape=<unknown> dtype=int32>

In [4]: tf.sparse_reduce_sum(x, axis=1, keepdims=True)
Out[4]: <tf.Tensor 'SparseReduceSum_1:0' shape=<unknown> dtype=int32>
```

**Will this change the current api? How?** 
No.
**Who will benefit with this feature?**
Users who use TensorFlow to do graph learning
**Any Other info.**

Hi, please check the following information, thanks!

Have I written custom code
N/A

OS Platform and Distribution
Windows 10

TensorFlow installed from
conda -c anaconda tensorflow-gpu

Bazel version
N/A

CUDA/cuDNN version
9.0/7.1.4

GPU model and memory
Quadro M520 2GB memory

Exact command to reproduce
describe in the original post

Mobile device
N/A

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #23114 where
the shape function of tf.sparse_reduce_sum did not
infer the shape even if the input shape were known.

This fix improves the shape function.

This fix fixes #23114.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/core/ops/sparse_ops.cc [hunk 1]
```diff
@@ -423,7 +423,45 @@ REGISTER_OP("SparseReduceSum")
     .Attr("keep_dims: bool = False")
     .Output("output: T")
     .Attr("T: numbertype")
-    .SetShapeFn(shape_inference::UnknownShape);
+    .SetShapeFn([](InferenceContext* c) {
```
Comment (akshaym) [inline tensorflow/core/ops/sparse_ops.cc]: Can this function be used for SparseReduceMax as well?

**Review by akshaym — COMMENTED** [review]
Please fix test: https://github.com/tensorflow/tensorflow/blob/f0dbcffb0eb44fb28d99a02905ac43e20ff7d5a2/tensorflow/core/ops/sparse_ops_test.cc#L134

**Comment by yongtang** [comment]
Thanks @akshaym, the PR has been updated with shape function applied to SparseReduceMax as well. Additional test cases have also been added. Please take a look and let me know if there are any issues.

File: tensorflow/core/ops/sparse_ops.cc [hunk 2]
```diff
@@ -39,6 +39,51 @@ Status SparseSparseMinOrMaxShapeFn(InferenceContext* c) {
   return Status::OK();
 }
 
+Status SparseReduceShapeFn(InferenceContext* c) {
```
Comment (akshaym) [inline tensorflow/core/ops/sparse_ops.cc]: This is a super nit, but would you consider putting it here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/common_shape_fns.cc

Comment (yongtang) [reply tensorflow/core/ops/sparse_ops.cc]: @akshaym Thanks for the suggestion. The shape function has been moved to `common_shape_fns.cc`.

**Comment by yongtang** [comment]
Thanks @akshaym for the review. The PR has been updated. Please take a look and let me know if there are any issues.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @akshaym: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by akshaym — APPROVED** [review]

**Pull request merged**